Java: Promote Insecure Basic Authentication query from experimental

atorralba opened the pull request:
PR to promote the Insecure Basic Authentication query created in #3976

## Changes
* Existing files were moved out of experimental.
* The `InsecureBasicAuth.qll` file was created.
* Refactored tests to use InlineExpectationsTest.

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-522/InsecureBasicAuth.ql:
```diff
@@ -0,0 +1,36 @@
+ * @precision medium
+ * @id java/insecure-basic-auth
+ * @tags security
+ *       external/cwe-522
+ *       external/cwe-319
```
```suggestion
 *       external/cwe/cwe-522
 *       external/cwe/cwe-319
```

intrigus-lgtm commented:
https://github.com/github/codeql/blob/2802736af7b164103276d0f2c9a075ff1adfe151/java/ql/src/Security/CWE/CWE-522/InsecureBasicAuth.qhelp#L18-L20
This should be removable, once the external references in the `.ql` file are fixed I think.

aschackmull commented on java/change-notes/2021-06-01-insecure-basic-auth-query.md:
```diff
@@ -0,0 +1,2 @@
+lgtm,codescanning
+* The query "Insecure basic authentication" (`java/insecure-basic-auth`) has been promoted from experimental to the main query pack. Its results will now appear by default. This query was originally [submitted as an experimental query by @luchua-bc](https://github.com/github/codeql/pull/3976)
```
```suggestion
* The query "Insecure basic authentication" (`java/insecure-basic-auth`) has been promoted from experimental to the main query pack. Its results will now appear by default. This query was originally [submitted as an experimental query by @luchua-bc](https://github.com/github/codeql/pull/3976).
```

atorralba commented:
@github/docs-content-codeql please review the `qhelp` file. Even though changes aren't introduced in this PR, it wasn't reviewed when this query was merged to experimental.

mchammer01 reviewed (commented):
@atorralba - looking good ✨ 
A couple of comments to do with the same sentence (`Transmission of sensitive information not in HTTPS is vulnerable to packet sniffing`), used in the ql and the qhelp file
- for the ql file, see my inline comment
- for the qhelp file, the sentence can be found here: https://github.com/github/codeql/pull/5983/files#diff-6cd7c3b725b0f2ab0b30bf1926faa89b1e357682c8871c32453481c1faa74fa1R5

Would be great if this could be reworded so that it reads better 🙂

atorralba commented on java/ql/src/Security/CWE/CWE-522/InsecureBasicAuth.ql:
```diff
@@ -0,0 +1,23 @@
+ * @name Insecure basic authentication
+ * @description Basic authentication only obfuscates username/password in
+ *              Base64 encoding, which can be easily recognized and reversed.
+ *              Transmission of sensitive information not over HTTPS is
```
Thanks for the review and the suggestion! It makes total sense from a technical point of view too :-) 

Fixed in 006c69bb85ab21eb132303dd0d9fb73cd0f4fc5d

aschackmull commented:
The example `InsecureBasicAuth.java` in the qhelp has some scope issues, so it could never compile regardless of context. That should be fixed.

aschackmull commented:
How much does this query overlap with the existing `HttpsUrls.ql`? Should they be combined into one query? At the very least it seems like they ought to share more code than they currently do.

atorralba commented:
Rebased main to solve conflicts and force-pushed.

Thanks for your review @aschackmull, and apologies for the late response. About your comments:

> The example InsecureBasicAuth.java in the qhelp has some scope issues, so it could never compile regardless of context. That should be fixed.

Fixed in c3c7337.

> How much does this query overlap with the existing HttpsUrls.ql? Should they be combined into one query? At the very least it seems like they ought to share more code than they currently do.

That's good insight. I refactored both queries and indeed a lot of code could be shared (I even managed to simplify some things using `localExprTaint` instead of convoluted string concatenation logic). All the changes to the query `HttpsUrls` are in 2e08c5d. Then, in 5ed9949 I use those changes to simplify `InsecureBasicAuth`.

I still think these two make sense as separate queries, since `HttpsUrls` is a `recommendation` and `InsecureBasicAuth` is a `warning`.
